File: 0109-convert-sorted-list-to-binary-search-tree/0109-convert-sorted-list-to-binary-search-tree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedListToBST(ListNode* head) {
        if (head==NULL) return NULL;
        return build(head,NULL);
    }
    TreeNode* build(ListNode* head,ListNode* tail){
        if (head==tail) return NULL;
        ListNode* fast = head;
        ListNode* slow = head;
        
        while (fast!=tail and fast->next!=tail){
            fast = fast->next->next;
            slow = slow->next;
        }
        TreeNode *root = new TreeNode(slow->val);
        root->left = build(head,slow);
        root->right = build(slow->next,tail);
        return root;
    }
};

File: 0109-convert-sorted-list-to-binary-search-tree/0109-convert-sorted-list-to-binary-search-tree.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedListToBST(self, head: Optional[ListNode]) -> Optional[TreeNode]:
        arr = []
        
        def inorder(node):
            while node:
                arr.append(node.val)
                node = node.next
        
        inorder(head)
        print(arr)
        
        
        def convert(start,end):
            if start>end:
                return None
            
            mid = start + (end-start)//2
            print(mid)
            root = TreeNode(arr[mid])
            root.left = convert(start,mid-1)
            root.right = convert(mid+1,end)
            
            
            return root
        
        return convert(0,len(arr)-1)
            
            
            
            
            
            
            